src/gui/guiFormSpecMenu.cpp

```diff
@@ -2695,6 +2696,83 @@ void GUIFormSpecMenu::parseSetFocus(const std::string &element)
 		<< "'" << std::endl;
 }
 
+void GUIFormSpecMenu::parseModel(parserData *data, const std::string &element)
+{
+	std::vector<std::string> parts = split(element, ';');
+
+	if (parts.size() < 4 || (parts.size() > 7 &&
+			m_formspec_version <= FORMSPEC_API_VERSION)) {
+		errorstream << "Invalid model element (" << parts.size() << "): '" << element
+			<< "'" << std::endl;
+		return;
+	}
+
+	parts.resize(std::max((s32)parts.size(), 7));
+
+	std::vector<std::string> v_pos = split(parts[0], ',');
+	std::vector<std::string> v_geom = split(parts[1], ',');
+	std::string meshstr = unescape_string(parts[2]);
+	std::vector<std::string> textures = split(parts[3], ',');
+	std::vector<std::string> vec_rot = split(parts[4], ',');
+	bool inf_rotation = is_yes(parts[5]);
+	bool mousectrl = is_yes(parts[6]) || parts[6].empty(); // default true
+
+	MY_CHECKPOS("model", 0);
+	MY_CHECKGEOM("model", 1);
+
+	v2s32 pos;
+	v2s32 geom;
+
+	if (data->real_coordinates) {
+		pos = getRealCoordinateBasePos(v_pos);
+		geom = getRealCoordinateGeometry(v_geom);
+	} else {
+		pos = getElementBasePos(&v_pos);
+		geom.X = stof(v_geom[0]) * (float)imgsize.X;
+		geom.Y = stof(v_geom[1]) * (float)imgsize.Y;
+	}
+
+	if (!data->explicit_size)
+		warningstream << "invalid use of model without a size[] element" << std::endl;
+
+	scene::IAnimatedMesh *mesh = m_client->getMesh(meshstr);
+
+	if (!mesh) {
+		errorstream << "Invalid model element: Unable to load mesh:"
+				<< std::endl << "\t" << meshstr << std::endl;
+		return;
+	}
+
+	FieldSpec spec(
+		meshstr,
+		L"",
+		L"",
+		258 + m_fields.size()
+	);
+
+	core::rect<s32> rect(pos, pos + geom);
+
+	GUIScene *e = new GUIScene(Environment, RenderingEngine::get_scene_manager(),
+			data->current_parent, rect, spec.fid);
+
+	auto meshnode = e->setMesh(mesh);
+
+	for (u32 i = 0; i < textures.size() && i < meshnode->getMaterialCount(); ++i)
+		e->setTexture(i, m_tsrc->getTexture(textures[i]));
+
+	if (vec_rot.size() >= 2)
```

Comment:
```suggestion
	if (vec_rot.size() == 2)
```
The rotation vector can have only two angles as it is 2d.

Reply:
Using `==` makes formspec ignore rotation if there are extra angles.
Using `>=` makes formspec ignore extra angles but still apply rotation. 
It may be better with `>=` for future backward compatibility if there are some additions. Not sure what to choose here.

Reply:
Ah, yes. The rotation should be applied in any case if there are two angles passed at least.